### PR TITLE
New FIPS break tests for ECDH and AES-GCM-256

### PR DESCRIPTION
**Pull request by samuel40791765**
### Issues:
Addresses `CryptoAlg-1739`

### Description of changes: 
Added new break tests for ECDH and AES-GCM-256. 

### Call-outs:
Since the ECDH functions call the EC self tests before running themselves, we need a special function called `ecdh_compute_shared_secret_no_self_test` to avoid an infinity loop. This follows the previous pattern for other EC self tests and shouldn't effect the original behavior of the existing ECDH APIs. The "no_self_test" function will be left internal like the other no_self_test functions.

### Testing:
Running the break test script with `./util/fipstools/test-break-kat.sh` passes the break testing.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and 
the ISC license.

### CODE REVIEW

**Comment by samuel40791765** [comment]
This seemed like a change that could go onto mainline, so I've opened it there first. We can cherrypick into onto `experimental-fips-refactoring` once it's approved.
I'll can reopen this if we feel like we should put this onto `experimental-fips-refactoring` first.

File: crypto/fipsmodule/self_check/self_check.c [hunk 1]
```diff
@@ -509,6 +509,88 @@ static int boringssl_self_test_rsa(void) {
...
+      0xa3, 0x8d, 0x12, 0xb7, 0x8c, 0xb1, 0x21, 0x89, 0xec, 0x67, 0x9d, 0xdd,
+      0x2e, 0x9a, 0xa2, 0x1f, 0xa9, 0xa5, 0xa0, 0xb0,
+  };
+
+  EC_KEY *key = EC_KEY_new();
+  EC_GROUP *group = EC_GROUP_new_by_curve_name(NID_secp384r1);
+  EC_POINT *pub_key = EC_POINT_new(group);
+  EC_POINT *peer_pub_key = EC_POINT_new(group);
+
+  BIGNUM *qx = BN_bin2bn(kQx, sizeof(kQx), NULL);
```
Comment (torben-hansen) [inline crypto/fipsmodule/self_check/self_check.c:555]: nit: you don't actually need all this intermediate stuff if you encode the coordinates as an EC point.

File: crypto/fipsmodule/ecdh/ecdh.c [hunk 2]
```diff
@@ -131,6 +130,15 @@ int ECDH_compute_shared_secret(uint8_t *buf, size_t *buflen, const EC_POINT *pub
   return ret;
 }
 
+int ECDH_compute_shared_secret(uint8_t *buf, size_t *buflen,
+                               const EC_POINT *pub_key,
+                               const EC_KEY *priv_key) {
+  boringssl_ensure_ecc_self_test();
+
+  return ecdh_compute_shared_secret_no_self_test(buf, buflen, pub_key,
```
Comment (torben-hansen) [inline crypto/fipsmodule/ecdh/ecdh.c:138]: OK, but we could prevent the same thread from entering `run_self_test_ecc` twice I guess. And we wouldn't need two functions?

**Review by torben-hansen — APPROVED** [review]

File: crypto/fipsmodule/ecdh/ecdh.c (hunk 2 above)
Comment (samuel40791765) [reply crypto/fipsmodule/ecdh/ecdh.c:138]: I'll keep this for now, so we're following our original pattern.
I'm curious though, is there a way to do that? Seems like `CRYPTO_once` only blocks concurrent threads from accessing the original executing thread, is there something that would help us block the original thread execution as well

Comment (torben-hansen) [reply crypto/fipsmodule/ecdh/ecdh.c:138]: yeah, just a thread-local semaphore

**Review by torben-hansen — APPROVED** [review]

**Comment by samuel40791765** [comment]
Closing, good news is that this isn't required anymore

**Pull request closed without merging**